### DISCUSSION TITLE
How to `usePrepareContractWrite` with args that are not ready at run time?

### DISCUSSION THREAD

**Discussion by alexiskattan** [wevm/wagmi#996] [Ideas]
I really love the library. Am having an issue, before we do the write, we first have to:

1) send arg to IPFS and get back the hash
2) put hash in args
3) set enable to true
```
const { config: contractWriteConfig } = usePrepareContractWrite({
     ...contractConfig,
     functionName: 'mint',
     args: [ipfsHash],
     enabled: Boolean(ipfsHash),
});
```

There is a bit of a chicken and egg situation. If we wait till args is ready to go, enabled is false, and we can't kick off the process to hash it. I'm wondering what is the proper way to handle this with the library?

**Comment by masonchain** [comment]
I recently faced the exact same issue myself!

I'm assuming since you really do have to prepare upon completion of the IPFS upload and trigger the contract write upon response then there is not much you can do to faithfully prepare in a one click UX action.

@jxom hopped into my PR and recommended putting the actual write call in a useEffect in a similar fashion to: 

```
const { config: contractWriteConfig, isSuccess } = usePrepareContractWrite({
     ...contractConfig,
     functionName: 'mint',
     args: [ipfsHash],
     enabled: Boolean(ipfsHash),
});
const { write } = useContractWrite(contractWriteConfig);

useEffect(() => {
  if (isSuccess) {
       write?.()
    }
}, [isSuccess]);
```


Some deviations from my actual implementation (used writeAsync within a separate function called within useEffect), but it has worked flawlessly so far. Hope this helped!

**Reply by tmm** [reply]
Note: You likely do not want to trigger `write` in a side effect like this since [it won't work with deeplinking on iOS](https://wagmi.sh/docs/prepare-hooks/intro#ios-app-link-constraints).

**Reply by masonchain** [reply]
> Note: You likely do not want to trigger `write` in a side effect like this since [it won't work with deeplinking on iOS](https://wagmi.sh/docs/prepare-hooks/intro#ios-app-link-constraints).

Is there a recommended solution besides abstracting any kind of argument fetching (e.g. an IPFS hash) behind another button click?

**Comment by alexiskattan** [comment]
I've been thinking about it.  The IPFS hash is of fixed size (but unknown value), so is there any way to use that bit of information to be able to precompute the gas fees at all and save the user time?

**Reply by masonchain** [reply]
That's a good question that may be better answered by someone more in touch with the inner workings. But, I would assume that even if you used a dummy string for the preparation that it would still be a brand new transaction being populated and estimated when the real ipfs hash is received.

**Comment by rsodre** [comment]
It is not clear on the docs if when `enabled` is true it will **actually write** automatically, or we're just saying that all the args are ready and it can prepare and wait for `write()` to be called (I'm counting on the latter)
